File: app/src/main/java/pe/kr/crasy/dunningassignments/Edit/EditPresenter.java
package pe.kr.crasy.dunningassignments.Edit;

import java.util.Date;

import io.realm.Realm;
import pe.kr.crasy.dunningassignments.database.Assignments;

/**
 *
 * Created by maybe on 16. 12. 29.
 */

public class EditPresenter {
    private Realm realm;
    private PView view;
    private EditModel model;
    public EditPresenter(PView view){
        this.view = view;
        this.model = new EditModel();
        this.realm = Realm.getDefaultInstance();
    }

    public void onTitleUpdate(String title){
        model.setTitle(title);
    }

    public void onLocationUpdate(String location){
        model.setLocation(location);
    }

    public void onMetaDataUpdate(String metaData){
        model.setMetaData(metaData);
    }

    public void closeActivity(){
        if(!isValid()){
            view.onCloseActivity(false);
            return;
        }
        realm.executeTransaction(realm1 -> {
            Assignments assignments = realm1.createObject(Assignments.class);
            assignments.setTitle(model.getTitle());
            assignments.setLocation(model.getLocation());
            assignments.setStartLine(model.getStartLine());
            assignments.setDeadLine(model.getDeadLine());
            assignments.setMetaData(model.getMetaData());
        });
        realm.close();
        view.onCloseActivity(true);
    }

    public void onDateUpdate(Date date, String title){
        if(title.equals("startLine")){
            model.setStartLine(date);
            view.onStartLineText(date);
        } else {
            model.setDeadLine(date);
            view.onDeadLineText(date);
        }
    }

    public boolean isValid(){
        return model.getTitle() != null && model.getLocation() != null && model.getDeadLine() != null;
    }

    public interface PView {
        void onCloseActivity(boolean isSaved);
        void onStartLineText(Date date);
        void onDeadLineText(Date date);
    }
}

File: app/src/main/java/pe/kr/crasy/dunningassignments/edit/EditPresenter.kt
package pe.kr.crasy.dunningassignments.edit

import java.util.ArrayList
import java.util.Date

import io.realm.Realm
import io.realm.RealmList
import pe.kr.crasy.dunningassignments.database.Assignments
import pe.kr.crasy.dunningassignments.database.People

/**

 * Created by maybe on 16. 12. 29.
 */

class EditPresenter(private val view: EditPresenter.PView) {
    private val realm: Realm = Realm.getDefaultInstance()
    private val model: EditModel = EditModel()

    fun onTitleUpdate(title: String) {
        model.title = title
    }

    fun onLocationUpdate(location: String) {
        model.location = location
    }

    fun onMetaDataUpdate(metaData: String) {
        model.metaData = metaData
    }

    fun closeActivity() {
        if (!isValid) {
            view.onCloseActivity(false)
            return
        }
        realm.executeTransaction { realm1 ->
            val assignments = realm1.createObject(Assignments::class.java)
            assignments.title = model.title
            assignments.location = model.location
            assignments.startLine = model.startLine
            assignments.deadLine = model.deadLine
            assignments.metaData = model.metaData

            val peoples = RealmList<People>()
            for (i in people.indices) {
                val person = People()
                person.name = people[i]
                peoples.add(person)
            }
            assignments.people = peoples
        }
        realm.close()
        view.onCloseActivity(true)
    }

    fun onDateUpdate(date: Date, title: String) {
        if (title == "startLine") {
            model.startLine = date
            view.onStartLineText(date)
        } else {
            model.deadLine = date
            view.onDeadLineText(date)
        }
    }

    private val people = ArrayList<String>()

    fun onNewMember(memberName: String, position: Int) {
        if (position >= people.size) {
            people.add(memberName)
        } else {
            people[position] = memberName
        }
        model.people = people
    }

    val isValid: Boolean
        get() = model.title != null && model.location != null && model.deadLine != null

    interface PView {
        fun onCloseActivity(isSaved: Boolean)
        fun onStartLineText(date: Date)
        fun onDeadLineText(date: Date)
    }
}
